### ISSUE TITLE
Spanner: could Database#client be an alternate way of getting a client?

### ISSUE THREAD

**Issue by remi** [GoogleCloudPlatform/google-cloud-ruby#1596]
Based on existing idioms used throughout the Google Cloud Client libraries for Ruby, I presumed that the following would work:

```ruby
require "google/cloud/spanner"

spanner  = Google::Cloud::Spanner.new
instance = spanner.instance  "my-instance"
database = instance.database "my-database"

client = database.client
client = instance.client "my-database"
client = spanner.client  "my-instance", "my-database"
```

Today, you can get a Database via either the Project or the Instance

```ruby
database = spanner.database  "my-instance", "my-database"
database = instance.database "my-database"
```

And similar idioms are used throughout google-cloud-ruby to get resource object.

tl;dr I found it surprising that database.client wasn't available and I have to pass the instance_id and database_id to Project#client, even thought I have Database/Instance objects that know these values.

I wonder (1) was this considered and why wasn't the interface implemented this way (2) could this possibly be an additive change? Thanks!

**Comment by blowmage** [comment]
It wasn't implemented that way because Spanner is really a combination of 3 different APIs:

* Data
* Instance
* Database

The Client uses the Data API, and is the real star of the library. We expect most users will focus on the Data API. `Project#instances` and `Project#instance` and `Instance` use the Instance API. `Project#databases`, `Project#database`, `Instance#databases`, `Instance#database`, and `Database` use the Database API. Its possible to have access to the Database API, but not have access to the Data API or the Instance API, so we needed a way for those users to access the Database API without having to access the other APIs. That is why `Project#databases` and `Project#database` exist.

Creating a Client from an Instance or Database object is possible, but I don't see much of a benefit. It seems overly inefficient to call the Instance API and then the Database API to create a Client to access the Data API. Again, its possible to do, but not something I expect we would want to encourage users to do.

**Comment by remi** [comment]
> Its possible to have access to the Database API, but not have access to the Data API or the Instance API, so we needed a way for those users to access the Database API without having to access the other APIs.

*AH* that makes perfect sense.  From an implementation point of view & helping users from shooting themselves in the foot, I totally understand.

As a new user using the library *(only considering the library interface)*, I found it surprising *(and a bit silly)* that my Database object couldn't give me a Client.

```ruby
client = spanner.client @instance.instance_id, @database.database_id
```

**Me:** *"Seriously?!"*

But... if my application were only given Data API access to read/write, then I could end up confused/frustrated about why I can't fetch Instance or Database objects so I can get my Database#client. So we wouldn't want our samples to show people code that will *absolutely* break for many applications. Most apps will never fetch Instance or Database objects :)

👌 Sounds good to me, thank you for the explanation!! I retract my suggestion 😊

**Comment by quartzmo** [comment]
OK to close this issue?
